src/object.d

```diff
@@ -3514,7 +3514,15 @@ private U[] _dup(T, U)(T[] a) if (!__traits(isPOD, T))
     import core.lifetime: copyEmplace;
     U[] res = () @trusted {
         auto arr = cast(U*) _d_newarrayU(typeid(T[]), a.length);
-        foreach (i; 0..a.length)
+        size_t i;
+        scope (failure)
+        {
+            import core.internal.lifetime: emplaceInitializer;
+            // Initialize all remaining elements to not destruct garbage
+            foreach (j; i .. a.length)
+                emplaceInitializer(cast() arr[j]);
```

Comment:
So you're resetting the element whose postblit/cctor threw - should we?

Reply:
Yes, I think it's necessary because the GC will call the destructor for that element.

Reply:
My concern was that the postblit/cpctor might not clean up itself on an exception and depend on the dtor being run on the partially constructed instance. But that's hopefully handled by `-preview=dtorfields`, and resetting to `T.init` correct then. Maybe slightly adjust the comment to make that clear, to prevent someone from thinking it's a one-off bug as I did. ;)

Reply:
Isn't the correct thing instead for `copyEmplace` to reset to `T.init` on failure, then this loop would be from `i + 1`?

Otherwise this fix needs to be done in a bunch of other places because it would mean `copyEmplace` always needs a `try`/`catch` or `scope(failure)` to use safely if it might throw.

Reply:
Yes, that seems to make a lot of sense. After all, regular instances failing their construction aren't destructed, so postblit/cpctor cannot rely on the dtor being run in case they throw. - Maybe just using a `try / catch(Exception)` to prevent the overhead for `nothrow` postblits and cpctors.

Reply:
=> https://issues.dlang.org/show_bug.cgi?id=22177